### PR TITLE
New events listing

### PR DESCRIPTION
**Pull request by peteryates**
### Trello card

https://trello.com/c/kzd6I5IZ/1588-events-improve-events-listing-page-and-details-page-designs

### Context

This PR contains a new (alternative) events listing. Instead of categorising by type the events will be in a single list with filters. To highlight 'Train to teach' events, the first two will be placed in a special box at the top of the results listing.

### Preview

Screenshot from 2021-10-29 17-11-23

Screenshot from 2021-10-29 16-28-29

### CODE REVIEW

**Comment by gemmadallmandfe** [comment]
@peteryates Silktide is flagging that we are not marking up our event listings as list items on the current page (https://index.silktide.com/uk-central-government/get-into-teaching/recommendations/mark-navigation-section-as-list).

Looks like this is not a problem on the new design but thought I should flag with you just in case.

**Comment by peteryates** [comment]
@gemmadallmandfe - yep, all the event lists are ordered lists.

Some previews of the WIP event page (no slantly lines yet!)

| Mobile | Desktop |
| -------- | -----------|
|Screenshot from 2021-11-01 16-37-31 | Screenshot from 2021-11-01 16-37-24 |

**Comment by peteryates** [comment]
Slow going on this one, design is really fiddly 😩 

Screenshot from 2021-11-02 16-47-46

File: app/views/teaching_events/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+          <% if @event.is_virtual %>
+            <div>
+              Virtual
+            </div>
+          <% end %>
+        </div>
+
+        <div class="register">
+          <%= link_to event_steps_path(@event.readable_id), class: %w[button button--black-chevron] do %>
+            Register your attendence
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show.html.erb:40]:
```suggestion
            Register your attendance
```

File: app/views/teaching_events/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+      <div class="triangle yellow">
+        <svg height="100%" width="100%" viewbox="0 0 100 100" preserveAspectRatio="none">
+          <polygon points="0,100 0,0 100,0" />
+        </svg>
+      </div>
+    </section>
+  </header>
+
+  <section class="event-info">
+    <h2>What to expect from a <%= event_type_name(@event.type_id) %> event</h2>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show.html.erb:59]: This says 'What to expect from a Train to Teach event **event**'

File: app/views/teaching_events/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    <%#   <li><%= @event.name %1></li> %>
+    <%#   <li><%= @event.summary %1></li> %>
+    <%#   <li>online: <%= @event.is_online %1></li> %>
+    <%#   <li>virtual: <%= @event.is_virtual %1></li> %>
+    <%#   <li><%= @event.building&.venue %1></li> %>
+    <%#   <li><%= @event.building&.address_postcode %1></li> %>
+    <%# </ul> %>
+  </section>
+
+  <section class="event-attributes">
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show.html.erb:87]: I'm not convinced by this section - I think it is a lot of repeated information that they already know from earlier up the page; takes up a lot of real estate!

File: app/models/teaching_events/search.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  private
+
+    def query(limit: 12)
+      conditions = {
+        type_ids: type_condition,
+        postcode: postcode_condition,
+        online: online_condition,
+        quantity_per_type: limit,
+        radius: distance_condition,
```
Comment (gemmadallmandfe) [inline app/models/teaching_events/search.rb:27]: Following conversation with the Events team, please could we change the radius as follows:

- remove 25 miles
- add 30 miles (this is what they use for their event comms)
- add 50 miles (as a test to see if people want a wider area)

File: app/views/teaching_events/_filter.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  <div class="events__filter--group">
+    <%= f.govuk_check_boxes_fieldset(:online, legend: nil) do %>
+      <%= f.govuk_check_box :online, "true", label: { text: "Online" } %>
+      <%= f.govuk_check_box :online, "false", label: { text: "In person" } %>
+    <% end %>
+  </div>
+
+  <div class="events__filter--group">
+    <%= f.govuk_check_boxes_fieldset(:type, legend: { text: "Event type", tag: nil }) do %>
+      <%= f.govuk_check_box :type, "222_750_001,222_750_007", label: { text: "Train to teach" } %>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/_filter.html.erb:20]:
```suggestion
      <%= f.govuk_check_box :type, "222_750_001,222_750_007", label: { text: "Train to Teach" } %>
```
Capitalisation as this is essentially a brand name

File: app/views/teaching_events/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    <% if @featured_events.blank? && @events.blank? %>
+      <div role="alert">
+        No events were found.
+      </div>
+    <% else %>
+      <div class="teaching-events__listing">
+        <% if @featured_events.any? %>
+          <div class="teaching-events__events teaching-events__events--featured">
+            <h2>Upcoming train to teach events</h2>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/index.html.erb:17]:
```suggestion
            <h2>Upcoming Train to Teach events</h2>
```

File: app/views/teaching_events/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,35 @@
+<section class="teaching-events">
+  <h1>Teaching events</h1>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/index.html.erb:2]:
```suggestion
  <h1>Get into Teaching events</h1>
```
We know from UR that people find 'Teaching events' hard to understand - are they for teachers?

Suggest we try this instead for now.

Should/could this be styled in the same way that the blog h1 is styled (numberplate!)?

File: app/views/teaching_events/show/_train-to-teach-event-summary.html.erb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<div class="about-ttt-event">
+  <p>Watch presentations, ask questions, speak to advisers one-to-one.</p>
+</div>
+
+<figure>
+  <blockquote>
+    Absolutely loved this day. Very informative, inspiring and I'm
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show/_train-to-teach-event-summary.html.erb:7]: Please can you change this quote to:

It was exactly what I needed to give me guidance on the next steps to be taken

File: app/views/teaching_events/show/_train-to-teach-event-summary.html.erb [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+    excited to change career. Thank you.
+  </blockquote>
+  <figcaption>Sarah, trainee teacher, London</figcaption>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show/_train-to-teach-event-summary.html.erb:10]: Can you delete the caption - we don't have names to use, so just going with the quotes

File: app/views/teaching_events/show/_what-theyre-saying.html.erb [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+    <svg height="100%" width="100%" viewbox="0 0 100 100" preserveAspectRatio="none">
+      <polygon points="0,100 100,0 100,100" />
+    </svg>
+  </div>
+  <div class="subheading">
+    <h3>What they're saying</h3>
+  </div>
+  <figure>
+    <blockquote>
+      Absolutely loved this day. Very informative, inspiring and I'm
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show/_what-theyre-saying.html.erb:12]: Please can you replace this quote with

So useful! I got answers to questions I didn't know I had yet and I'm so inspired and excited.

File: app/views/teaching_events/show/_what-theyre-saying.html.erb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+      excited to change career. Thank you.
+    </blockquote>
+    <figcaption>Sarah, trainee teacher, London</figcaption>
```
Comment (gemmadallmandfe) [inline app/views/teaching_events/show/_what-theyre-saying.html.erb:15]: Please can you remove the caption - we are not including names

File: app/views/teaching_events/show/_event-attributes.html.erb [hunk 12]
```diff
@@ -0,0 +1,25 @@
+<section class="event-attributes">
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/views/teaching_events/show/_event-attributes.html.erb:1]: Ok, I'll remove this

File: config/initializers/date_formats.rb [hunk 13]
```diff
@@ -8,3 +8,4 @@
 Time::DATE_FORMATS[:time] = "%H:%M"
 Date::DATE_FORMATS[:govuk_date_long] = "%A %-d %B"
```
[the remark this answers is not in the document]
Comment (peteryates) [reply config/initializers/date_formats.rb:9]: Fixed in e903db15dc

File: app/components/teaching_events/event_component.html.erb [hunk 14]
```diff
@@ -0,0 +1,59 @@
+<%= tag.li(class: classes) %>
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/components/teaching_events/event_component.html.erb:1]: I've done the TTT one, how do we know which non-TTT ones get a DFE stamp?

[the remark this answers is not in the document]
Comment (gemmadallmandfe) [reply app/components/teaching_events/event_component.html.erb:1]: Online forums should all be DfE? It is only School and University events that are third party provider events

(NB the data in the design screenshot above is wrong - Stour Vale online would be a School or University event, not an Online forum)

File: app/views/teaching_events/index.html.erb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
```
[the remark this answers is not in the document]
Comment (gemmadallmandfe) [reply app/views/teaching_events/index.html.erb:3]: Copy from Tom:

"All events listed here are designed to help you get into teacher training. They are all free to attend."

[the remark this answers is not in the document]
Comment (peteryates) [reply app/views/teaching_events/index.html.erb:3]: 0af5e5e6c080d80

File: app/components/teaching_events/event_component.rb [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    def train_to_teach?
+      type.in?([222_750_001, 222_750_007])
+    end
+
+    def school_and_university?
+      type == 222_750_009
+    end
+
+    def online?
+      @online || @virtual
```
Comment (ethax-ross) [inline app/components/teaching_events/event_component.rb:29]: All virtual events should have `is_online` as `true` as well, so you could get away with just `@online` here I think

File: app/components/teaching_events/event_component.rb [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+      @start_at  = event.start_at
+      @end_at    = event.end_at
+      @slug      = event.readable_id
+
+      super
+    end
+
+    # check if event is 'train to teach' or 'question time'
+    def train_to_teach?
+      type.in?([222_750_001, 222_750_007])
```
Comment (ethax-ross) [inline app/components/teaching_events/event_component.rb:21]: Worth references the constants here and other places they are hard coded?

https://github.com/DFE-Digital/get-into-teaching-app/blob/master/lib/extensions/get_into_teaching_api_client/constants.rb#L3-L9

File: app/controllers/teaching_events_controller.rb [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+private
+
+  def search_params
+    params.permit(teaching_events_search: [:postcode, :distance, { online: [], type: [] }])[:teaching_events_search]
+  end
+
+  def setup_filter
+    @distances = [
```
Comment (ethax-ross) [inline app/controllers/teaching_events_controller.rb:49]: Would it make more sense for this to be in the search model?

File: spec/components/teaching_events/event_component_spec.rb [hunk 19]
```diff
@@ -0,0 +1,193 @@
...
+
+        it { expect(subject).not_to be_in_person }
+      end
+    end
+
+    describe "#in_person?" do
+      context "when the event is a school and university event" do
+        let(:event) { build(:event_api, :virtual) }
+
+        it { expect(subject).to be_online }
```
Comment (ethax-ross) [inline spec/components/teaching_events/event_component_spec.rb:73]: Not sure this test makes sense? should it not construct an event that's in person then check `be_in_person`?

File: spec/components/teaching_events/event_component_spec.rb [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+      context "when the event has a location" do
+        let(:event) { build(:event_api) }
+
+        it { expect(subject).to be_in_person }
+      end
+
+      context "when the event has no location" do
+        let(:event) { build(:event_api, :no_location) }
+
+        it { expect(subject).not_to be_in_person }
```
Comment (ethax-ross) [inline spec/components/teaching_events/event_component_spec.rb:65]: would `be_online` read better here?

File: spec/helpers/teaching_events_helper_spec.rb [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+    specify "returns true when there's a match" do
+      expect(is_event_type?(ttt_event, ttt)).to be true
+    end
+
+    specify "returns false when there's no match" do
+      expect(is_event_type?(ttt_event, qt)).to be false
+    end
+  end
+
+  describe "#event_type_name" do
```
Comment (ethax-ross) [inline spec/helpers/teaching_events_helper_spec.rb:75]: Missing specs for `#is_a_train_to_teach_event`?

**Review by ethax-ross — COMMENTED** [review]
Looking good overall, just a couple comments/suggestions

File: spec/components/teaching_events/event_component_spec.rb (hunk 20 above)
Comment (peteryates) [reply spec/components/teaching_events/event_component_spec.rb:65]: Ah this is just checking whether the building's present, not that the event will definitely be either online or in person.

File: app/views/teaching_events/index/_filter.html.erb [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+    <% end %>
+  </div>
+
+  <div class="events__filter--group">
+    <%= f.govuk_check_boxes_fieldset(:type, legend: { text: "Event type", tag: nil }) do %>
+      <%= f.govuk_check_box :type, "222_750_001,222_750_007", label: { text: "Train to teach" } %>
+      <%= f.govuk_check_box :type, "222_750_008", label: { text: "Online forum" } %>
+      <%= f.govuk_check_box :type, "222_750_009", label: { text: "School and university" } %>
+    <% end %>
+  </div>
```
[the remark this answers is not in the document]
Comment (peteryates) [reply app/views/teaching_events/index/_filter.html.erb:24]: I'll do this as a subsequent PR

**Review by ethax-ross — APPROVED** [review]
Looks great 👍 after this rolls out we can maybe change the API endpoint to serve the events in a more suitable format. Only other thought I had was we limit to 12 events/type so ~30 events which won't lead to more than 1 page's worth... do we need to think about retrieving more now that we don't have the date filter so that people can jump ahead a few months potentially?

**Pull request merged**